File: 2026.5/3120. 统计特殊字母的数量 I/3120_special_chars.cpp
#include <iostream>
#include <string>
#include <unordered_set>
#include <vector>

using namespace std;

class Solution {
public:
    /**
     * 统计特殊字母的数量
     *
     * 特殊字母定义：如果 word 中同时存在某个字母的小写形式和大写形式，
     * 则称这个字母为特殊字母。
     *
     * @param word 输入字符串
     * @return 特殊字母的数量
     */
    int numberOfSpecialChars(string word) {
        // 使用 unordered_set 记录出现的小写字母和大写字母
        unordered_set<char> lower_set;
        unordered_set<char> upper_set;

        // 遍历字符串，统计大小写字母
        for (char c : word) {
            if (islower(c)) {
                lower_set.insert(c);
            } else if (isupper(c)) {
                upper_set.insert(c);
            }
        }

        // 统计同时存在大小写的字母
        int count = 0;
        for (char c : lower_set) {
            if (upper_set.count(toupper(c))) {
                count++;
            }
        }

        return count;
    }
};

// 测试函数
int main() {
    Solution solution;

    // 测试用例
    vector<pair<string, int>> test_cases = {
        {"aaAbcBC", 3},      // 特殊字母: a, b, c
        {"abc", 0},           // 无特殊字母
        {"ABC", 0},           // 无特殊字母
        {"aA", 1},            // 特殊字母: a
        {"", 0},              // 空字符串
        {"aAbBcCdD", 4},      // 特殊字母: a, b, c, d
    };

    // 运行测试
    for (const auto& test : test_cases) {
        string word = test.first;
        int expected = test.second;
        int result = solution.numberOfSpecialChars(word);

        string status = (result == expected) ? "PASS" : "FAIL";
        cout << "输入: '" << word << "' -> 输出: " << result
             << " (期望: " << expected << ") " << status << endl;
    }

    return 0;
}

File: 2026.05/3120. 统计特殊字母的数量 I/3120_special_chars.py
class Solution:
    def numberOfSpecialChars(self, word: str) -> int:
        """
        统计特殊字母的数量

        特殊字母定义：如果 word 中同时存在某个字母的小写形式和大写形式，
        则称这个字母为特殊字母。

        Args:
            word: 输入字符串

        Returns:
            特殊字母的数量
        """
        # 使用集合记录出现的小写字母和大写字母
        lower_set = set()
        upper_set = set()

        for char in word:
            if char.islower():
                lower_set.add(char)
            elif char.isupper():
                upper_set.add(char)

        # 统计同时存在大小写的字母
        count = 0
        for char in lower_set:
            if char.upper() in upper_set:
                count += 1

        return count

    def numberOfSpecialCharsConcise(self, word: str) -> int:
        """
        更简洁的实现方式

        Args:
            word: 输入字符串

        Returns:
            特殊字母的数量
        """
        lower = {c for c in word if c.islower()}
        upper = {c for c in word if c.isupper()}
        return len(lower & {c.lower() for c in upper})


# 测试代码
if __name__ == "__main__":
    solution = Solution()

    # 测试示例
    test_cases = [
        ("aaAbcBC", 3),      # 特殊字母: a, b, c
        ("abc", 0),           # 无特殊字母
        ("ABC", 0),           # 无特殊字母
        ("aA", 1),            # 特殊字母: a
        ("", 0),              # 空字符串
        ("aAbBcCdD", 4),      # 特殊字母: a, b, c, d
    ]

    for word, expected in test_cases:
        result = solution.numberOfSpecialChars(word)
        status = "PASS" if result == expected else "FAIL"
        print(f"输入: '{word}' -> 输出: {result} (期望: {expected}) {status}")
